File: frontend/lib/models/appointment.dart
/// Status of a booked appointment.
enum AppointmentStatus { upcoming, completed, cancelled }

extension AppointmentStatusX on AppointmentStatus {
  String get label {
    switch (this) {
      case AppointmentStatus.upcoming:
        return 'Upcoming';
      case AppointmentStatus.completed:
        return 'Completed';
      case AppointmentStatus.cancelled:
        return 'Cancelled';
    }
  }
}

/// Payment method used to confirm a booking.
enum PaymentMethod { card, upi, wallet }

extension PaymentMethodX on PaymentMethod {
  String get label {
    switch (this) {
      case PaymentMethod.card:
        return 'Credit / Debit Card';
      case PaymentMethod.upi:
        return 'UPI';
      case PaymentMethod.wallet:
        return 'Wallet';
    }
  }
}

/// A booked appointment with a doctor.
class Appointment {
  final String id;
  final String doctorId;
  final String doctorName;
  final String doctorPhotoUrl;
  final String specialtyName;
  final String hospitalName;
  final DateTime dateTime;
  final String slotLabel; // e.g. "10:30 AM"
  final int fee;
  final PaymentMethod paymentMethod;
  final String? patientName;
  final int? patientAge;
  final String? patientBloodGroup;
  final String? patientType;
  final String? paymentStatus;
  AppointmentStatus status;
  bool reviewed;

  Appointment({
    required this.id,
    required this.doctorId,
    required this.doctorName,
    required this.doctorPhotoUrl,
    required this.specialtyName,
    required this.hospitalName,
    required this.dateTime,
    required this.slotLabel,
    required this.fee,
    required this.paymentMethod,
    this.status = AppointmentStatus.upcoming,
    this.reviewed = false,
    this.patientName,
    this.patientAge,
    this.patientBloodGroup,
    this.patientType,
    this.paymentStatus,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'doctorId': doctorId,
        'doctorName': doctorName,
        'doctorPhotoUrl': doctorPhotoUrl,
        'specialtyName': specialtyName,
        'hospitalName': hospitalName,
        'dateTime': dateTime.toIso8601String(),
        'slotLabel': slotLabel,
        'fee': fee,
        'paymentMethod': paymentMethod.index,
        'status': status.index,
        'reviewed': reviewed,
        'patientName': patientName,
        'patientAge': patientAge,
        'patientBloodGroup': patientBloodGroup,
        'patientType': patientType,
        'paymentStatus': paymentStatus,
      };

  factory Appointment.fromJson(Map<String, dynamic> json) => Appointment(
        id: (json['id'] ?? json['_id'] ?? '') as String,
        doctorId: (json['doctorId'] ?? '') as String,
        doctorName: (json['doctorName'] ?? '') as String,
        doctorPhotoUrl: (json['doctorPhotoUrl'] ?? '') as String,
        specialtyName: (json['specialtyName'] ?? '') as String,
        hospitalName: (json['hospitalName'] ?? '') as String,
        dateTime: DateTime.parse(json['dateTime'] as String),
        slotLabel: (json['slotLabel'] ?? '') as String,
        fee: ((json['fee'] ?? 0) as num).toInt(),
        paymentMethod: PaymentMethod
            .values[((json['paymentMethod'] ?? 1) as int).clamp(0, 2)],
        status:
            AppointmentStatus.values[((json['status'] ?? 0) as int).clamp(0, 2)],
        reviewed: json['reviewed'] as bool? ?? false,
        patientName: json['patientName'] as String?,
        patientAge: json['patientAge'] as int?,
        patientBloodGroup: json['patientBloodGroup'] as String?,
        patientType: json['patientType'] as String?,
        paymentStatus: json['paymentStatus'] as String?,
      );
}

File: backend/src/models/Appointment.js
'use strict';

const crypto = require('crypto');
const mongoose = require('mongoose');
const {
  APPOINTMENT_STATUS,
  APPOINTMENT_STATUS_LABELS,
  PAYMENT_METHOD_LABELS,
  APPOINTMENT_SOURCE,
} = require('../constants');

// Backstop: anything more than this many seconds past its dateTime is auto-
// removed by MongoDB even if the hospital backend's end-of-day purge was missed.
// Future-dated appointments are never touched (TTL only fires once dateTime is
// in the past). Kept identical to the hospital backend on the shared collection.
const TTL_BACKSTOP_SECONDS = 3 * 24 * 60 * 60; // 3 days

/**
 * Appointment — the rolling-window store (today + future). Shape is a
 * byte-for-byte superset of the patient app's `Appointment.toJson()` so the app
 * can adopt this API unchanged: `status` and `paymentMethod` are stored as the
 * patient enum's integer index.
 *
 * `patientId` links the booking to the account that made it (so the app's "My
 * Appointments" returns only the signed-in patient's rows). It's optional so a
 * reception walk-in created via the hospital backend still validates on the
 * shared collection.
 */
const appointmentSchema = new mongoose.Schema(
  {
    _id: { type: String, default: () => `apt_${crypto.randomUUID()}` },

    // ── patient-app compatible fields ───────────────────────────────────────
    doctorId: { type: String, required: true },
    doctorName: { type: String, required: true },
    doctorPhotoUrl: { type: String, default: '' },
    specialtyName: { type: String, default: '' },
    hospitalName: { type: String, default: '' },
    dateTime: { type: Date, required: true },
    slotLabel: { type: String, required: true }, // "10:30 AM"
    fee: { type: Number, default: 0 },
    paymentMethod: { type: Number, default: 1 }, // 1 = UPI (patient enum index)
    status: {
      type: Number,
      enum: Object.values(APPOINTMENT_STATUS),
      default: APPOINTMENT_STATUS.UPCOMING,
    },
    reviewed: { type: Boolean, default: false },

    // ── ownership / desk metadata ───────────────────────────────────────────
    patientId: { type: String, index: true }, // app account that booked it
    patientName: { type: String },
    patientPhone: { type: String },
    patientAge: { type: Number },
    patientGender: { type: String, enum: ['male', 'female', 'other'] },
    patientBloodGroup: { type: String },
    patientType: { type: String, enum: ['ipd', 'opd'] },
    paymentStatus: { type: String, default: 'completed' },
    source: {
      type: String,
      enum: Object.values(APPOINTMENT_SOURCE),
      default: APPOINTMENT_SOURCE.PATIENT_APP,
    },
    checkedIn: { type: Boolean, default: false },
    tokenNumber: { type: Number },

    // Clinic-local calendar day (YYYY-MM-DD). Drives the rolling window and the
    // hospital backend's end-of-day pull/summarize/purge.
    dayKey: { type: String, required: true, index: true },
  },
  { timestamps: true, _id: false }
);

// "My appointments" queries (per patient, newest first).
appointmentSchema.index({ patientId: 1, dateTime: -1 });
// Availability / slot-conflict scans.
appointmentSchema.index({ doctorId: 1, dateTime: 1 });
// Hard double-booking guard: at most one UPCOMING appointment per
// doctor + day + slot. The app's isSlotBooked() is advisory; this is the truth.
appointmentSchema.index(
  { doctorId: 1, dayKey: 1, slotLabel: 1 },
  {
    unique: true,
    partialFilterExpression: { status: APPOINTMENT_STATUS.UPCOMING },
  }
);
// TTL backstop against a missed purge.
appointmentSchema.index({ dateTime: 1 }, { expireAfterSeconds: TTL_BACKSTOP_SECONDS });

// Emit the patient-compatible JSON shape: `id` (not `_id`), plus convenience
// string labels that the patient app harmlessly ignores.
appointmentSchema.set('toJSON', {
  transform(_doc, ret) {
    ret.id = ret._id;
    delete ret._id;
    delete ret.__v;
    ret.statusLabel = APPOINTMENT_STATUS_LABELS[ret.status];
    ret.paymentMethodLabel = PAYMENT_METHOD_LABELS[ret.paymentMethod];
    return ret;
  },
});

module.exports = mongoose.model('Appointment', appointmentSchema);
module.exports.TTL_BACKSTOP_SECONDS = TTL_BACKSTOP_SECONDS;
